### PR TITLE
[Serve] Json logging format

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

Update the logging format to json format. Used for better parsing and log search.

- User can set `RAY_SERVE_ENABLE_JSON_LOGGING` to jsonify the log message.
- Stream log doesn't have effect with this change.

controller log file

```
{"levelname": "INFO", "asctime": "2023-05-07 17:22:58,360", "component_name": "controller", "component_id": "3525674", "message": "http_state.py:129 - Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-9e1688af72409c6ffaf805b05c397632cfb4eb7acf0703468e8e3535' on node '9e1688af72409c6ffaf805b05c397632cfb4eb7acf0703468e8e3535' listening on '127.0.0.1:8000'"}
{"levelname": "INFO", "asctime": "2023-05-07 17:22:59,252", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1220 - Deploying new version of deployment app_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:22:59,282", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1459 - Adding 2 replicas to deployment app_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:22:59,283", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:330 - Starting replica app_testv2#QZdGDm for deployment app_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:22:59,297", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:330 - Starting replica app_testv2#jorUqs for deployment app_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,217", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1615 - Replica app_testv2#QZdGDm started successfully."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,217", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1615 - Replica app_testv2#jorUqs started successfully."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,270", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1220 - Deploying new version of deployment app2_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,319", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1459 - Adding 2 replicas to deployment app2_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,319", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:330 - Starting replica app2_testv2#lIYdIP for deployment app2_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,332", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:330 - Starting replica app2_testv2#FxSRtl for deployment app2_testv2."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:01,254", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1615 - Replica app2_testv2#lIYdIP started successfully."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:01,255", "component_name": "controller", "component_id": "3525674", "message": "deployment_state.py:1615 - Replica app2_testv2#FxSRtl started successfully."}
```

http proxy

```
{"levelname": "INFO", "asctime": "2023-05-07 17:22:59,242", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "message": "http_proxy.py:185 - Got updated endpoints: {}."}
{"levelname": "INFO", "asctime": "2023-05-07 17:22:59,253", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "message": "http_proxy.py:185 - Got updated endpoints: {'app_testv2': EndpointInfo(route='/app1', app_name='app')}."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:00,272", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "message": "http_proxy.py:185 - Got updated endpoints: {'app_testv2': EndpointInfo(route='/app1', app_name='app'), 'app2_testv2': EndpointInfo(route='/app2', app_name='app2')}."}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:06,895", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "request_id": "rzqadzRWFP", "route": "/app1", "app_name": "app", "message": "http_proxy.py:435 - GET 200 4.8ms"}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:08,168", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "request_id": "hYjyoiHTPJ", "route": "/app2", "app_name": "app2", "message": "http_proxy.py:435 - GET 200 4.6ms"}
{"levelname": "INFO", "asctime": "2023-05-07 17:23:32,596", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "message": "http_proxy.py:185 - Got updated endpoints: {'app_testv2': EndpointInfo(route='/app1', app_name='app'), 'app2_testv2': EndpointInfo(route='/app2', app_name='app2')}."}
{"levelname": "INFO", "asctime": "2023-05-07 17:24:02,716", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "message": "http_proxy.py:185 - Got updated endpoints: {'app_testv2': EndpointInfo(route='/app1', app_name='app'), 'app2_testv2': EndpointInfo(route='/app2', app_name='app2')}."}
{"levelname": "INFO", "asctime": "2023-05-07 17:24:35,044", "component_name": "http_proxy", "component_id": "b'172.31.5.229'", "message": "http_proxy.py:185 - Got updated endpoints: {'app_testv2': EndpointInfo(route='/app1', app_name='app'), 'app2_testv2': EndpointInfo(route='/app2', app_name='app2')}."}
```

deployment:

```
{"levelname": "INFO", "asctime": "2023-05-12 17:38:00,698", "deployment": "app2_Model", "replica": "app2_Model#XlFWYc", "request_id": "OfbLbdKgjT", "route": "/class_method", "application": "app2", "message": "replica.py:440 - Started executing request OfbLbdKgjT"}
{"levelname": "INFO", "asctime": "2023-05-12 17:38:00,699", "deployment": "app2_Model", "replica": "app2_Model#XlFWYc", "request_id": "OfbLbdKgjT", "route": "/class_method", "application": "app2", "message": "test_logging.py:200 - user log message from class method"}
{"levelname": "INFO", "asctime": "2023-05-12 17:38:00,699", "deployment": "app2_Model", "replica": "app2_Model#XlFWYc", "request_id": "OfbLbdKgjT", "route": "/class_method", "application": "app2", "message": "replica.py:537 - __CALL__ OK 0.6ms"}
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/logging_utils.py [hunk 1]
```diff
@@ -22,21 +22,27 @@ class ServeFormatter(logging.Formatter):
     """
 
     def __init__(self, component_name: str, component_id: str):
-        self.component_log_fmt = COMPONENT_LOG_FMT.format(
-            component_name=component_name, component_id=component_id
-        )
+        self.component_log_fmt = {
+            "levelname": "%(levelname)s",
+            "asctime": "%(asctime)s",
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/logging_utils.py:27]: Could you add time zone to the `asctime`?

**Review by edoakes — CHANGES_REQUESTED** [review]
I don't think it's an acceptable UX degradation to stream JSON-formatted logs to the console. We already have separate logging handlers for each, can we make the console logs human-readable (unchanged from today) and only log to JSON in the `serve/` log files?

Also, this needs to be guarded by a feature flag as it will likely break users' log parsers.

**Comment by sihanwang41** [comment]
> I don't think it's an acceptable UX degradation to stream JSON-formatted logs to the console. We already have separate logging handlers for each, can we make the console logs human-readable (unchanged from today) and only log to JSON in the `serve/` log files?
> 
> Also, this needs to be guarded by a feature flag as it will likely break users' log parsers.

Updated!

File: python/ray/serve/tests/test_logging.py [hunk 2]
```diff
@@ -156,9 +164,12 @@ def __call__(self, *args):
             assert replica_tag not in f.getvalue()
 
 
-def test_context_information_in_logging(serve_instance):
+@pytest.mark.parametrize("json_log_format", ["0", "1"])
```
Comment (edoakes) [inline python/ray/serve/tests/test_logging.py:167]:
```suggestion
@pytest.mark.parametrize("json_log_format", [True, False])
```
leaking implementation detail of the feature flag here

File: python/ray/serve/_private/logging_utils.py [hunk 3]
```diff
@@ -1,18 +1,60 @@
...
 LOG_FILE_FMT = "{component_name}_{component_id}.log"
-COMPONENT_LOG_FMT = (
-    "%(levelname)s %(asctime)s {component_name} {component_id} "  # noqa:E501
-)
 MESSAGE_FMT = "%(filename)s:%(lineno)d - %(message)s"
-REQUEST_ID_FMT = "%(request_id)s "
-ROUTE_FMT = "%(route)s "
+
+# Whether to jsonify the log message.
+FLAGS_JSONIFY_LOG_MESSAGE = os.environ.get(JSONIFY_LOG_MESSAGE, "0") == "1"
```
Comment (edoakes) [inline python/ray/serve/_private/logging_utils.py:19]: would prefer to define this directly in `constants.py` instead of adding the level of indirection

File: python/ray/serve/_private/logging_utils.py [hunk 4]
```diff
@@ -21,18 +63,27 @@ class ServeFormatter(logging.Formatter):
...
+        "%(levelname)s %(asctime)s {component_name} {component_id} "  # noqa:E501
+    )
+
     def __init__(self, component_name: str, component_id: str):
-        self.component_log_fmt = COMPONENT_LOG_FMT.format(
+        self.component_log_fmt = ServeFormatter.COMPONENT_LOG_FMT.format(
             component_name=component_name, component_id=component_id
         )
 
     def format(self, record):
```
Comment (edoakes) [inline python/ray/serve/_private/logging_utils.py:78]: can you type annotate `record` pls?

File: python/ray/serve/_private/logging_utils.py [hunk 5]
```diff
@@ -1,18 +1,60 @@
...
-)
 MESSAGE_FMT = "%(filename)s:%(lineno)d - %(message)s"
-REQUEST_ID_FMT = "%(request_id)s "
-ROUTE_FMT = "%(route)s "
+
+# Whether to jsonify the log message.
+FLAGS_JSONIFY_LOG_MESSAGE = os.environ.get(JSONIFY_LOG_MESSAGE, "0") == "1"
+
+
+class ServeJsonFormatter(logging.Formatter):
```
Comment (edoakes) [inline python/ray/serve/_private/logging_utils.py:22]:
```suggestion
class ServeJSONFormatter(logging.Formatter):
```

File: python/ray/serve/_private/logging_utils.py [hunk 6]
```diff
@@ -1,18 +1,60 @@
...
+
+    def format(self, record):
+        # generate a format string based on the record field.
+        cur_format = copy.deepcopy(self.component_log_fmt)
+        if "request_id" in record.__dict__:
+            cur_format["request_id"] = ServeJsonFormatter.REQUEST_ID_FMT
+        if "route" in record.__dict__:
+            cur_format["route"] = ServeJsonFormatter.ROUTE_FMT
+        if "app_name" in record.__dict__:
+            cur_format["app_name"] = ServeJsonFormatter.APP_NAME_FMT
```
Comment (edoakes) [inline python/ray/serve/_private/logging_utils.py:49]: I don't really understand how this works, will Python magically substitute in the values for these?

Comment (edoakes) [reply python/ray/serve/_private/logging_utils.py:49]: also, slight preference for `application` instead of `app_name`

**Review by edoakes — COMMENTED** [review]
I think it's worth writing some unit tests for the log formatters to ensure the format is stable over time and handles edge cases (like missing or extra fields)

File: python/ray/serve/_private/logging_utils.py (hunk 6 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py:49]: yes, it will automatically substitute the attribute values, similar to the `filename` and `lineno` etc.

File: python/ray/serve/tests/test_logging.py (hunk 2 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_logging.py:167]: good point.

**Comment by edoakes** [comment]
For deployment logs, is it possible to have `application: <name>`, `deployment: <name>`, `replica: <name>` instead of the "component" and "component_id" fields?

**Comment by jjyao** [comment]
cc @rkooo567, want to make sure this aligns with the overall logging improvement that Sang is working on.

File: python/ray/serve/_private/logging_utils.py [hunk 7]
```diff
@@ -1,18 +1,60 @@
...
+            "asctime": "%(asctime)s",
+            "component_name": component_name,
+            "component_id": component_id,
+        }
+
+    def format(self, record):
+        # generate a format string based on the record field.
+        cur_format = copy.deepcopy(self.component_log_fmt)
+        if "request_id" in record.__dict__:
+            cur_format["request_id"] = ServeJsonFormatter.REQUEST_ID_FMT
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py]: Make all string to constants

**Comment by rkooo567** [comment]
yeah this lgtm

File: python/ray/serve/_private/constants.py [hunk 8]
```diff
@@ -147,3 +147,27 @@ class ServeHandleType(str, Enum):
     "Please see the documentation for ServeDeploySchema for more details on multi-app "
     "config files."
 )
+
+# Jsonify the log messages
+SERVE_JSONIFY_LOG_MESSAGE = os.environ.get("SERVE_JSONIFY_LOG_MESSAGE") == "1"
```
Comment (edoakes) [inline python/ray/serve/_private/constants.py:152]:
```suggestion
# Jsonify the log messages
RAY_SERVE_ENABLE_JSON_LOGGING = os.environ.get("RAY_SERVE_ENABLE_JSON_LOGGING") == "1"
```

**Comment by sihanwang41** [comment]
ping ^^ @edoakes

**Comment by edoakes** [comment]
@sihanwang41 looks awesome, but I think we're missing `replica` in the deployment logs?
```
{"levelname": "INFO", "asctime": "2023-05-09 20:30:39,982", "deployment": "app2_Model", "request_id": "zhSPkLEaSF", "route": "/class_method", "application": "app2", "message": "test_logging.py:196 - user log message from class method"}
```

File: python/ray/serve/_private/http_state.py [hunk 9]
```diff
@@ -182,6 +182,13 @@ def _get_target_nodes(self) -> List[Tuple[str, str]]:
         location = self._config.location
         target_nodes = get_all_node_ids(self._gcs_client)
 
+        # Make ip address as string if it is bytes.
+        for index in range(len(target_nodes)):
+            node_id, ip_address = target_nodes[index]
+            if type(ip_address) == bytes:
+                ip_address = ip_address.decode("utf-8")
+                target_nodes[index] = (node_id, ip_address)
```
Comment (edoakes) [inline python/ray/serve/_private/http_state.py:190]: huh, is this change required for this PR? is it because we want to write the node_ip in string to the logs now?

**Review by edoakes — COMMENTED** [review]
Looks good, just would like to add `replica` to the logging format for deployments!

File: python/ray/serve/_private/http_state.py (hunk 9 above)
Comment (sihanwang41) [reply python/ray/serve/_private/http_state.py:190]: yes, otherwise json dumps will raise error.
Our function/method annotation says it is `str`, yet ip address is `bytes`.

Comment (zcin) [reply python/ray/serve/_private/http_state.py:190]: Ah I ran into the same issue, I added the bytes decoding [here](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/utils.py#L177). You should probably keep either my change or your change here.

Comment (sihanwang41) [reply python/ray/serve/_private/http_state.py:190]: nice!

Comment (edoakes) [reply python/ray/serve/_private/http_state.py:190]: prefer where cindy has it so it's fixed everywhere we use that util

Comment (sihanwang41) [reply python/ray/serve/_private/http_state.py]: rebased

**Review by edoakes — APPROVED** [review]
